qcog_python_client/qcog/qcogsync.py

```diff
@@ -0,0 +1,332 @@
+
+        seed: int, default=42
+            Seed
+
+        target_operator: list[str | int], default=None
+            Target operator
+
+
+        Returns
+        -------
+        QcogClient
+
+        """
+        super().pauli(*args, **kwargs)
+        return self
+
+    def ensemble(self, *args: Any, **kwargs: Any) -> QcogClient:  # noqa: D417
+        """Select Ensemble model.
+
+        Parameters
+        ----------
+        operators: list[str | int]
+            List of operators
+
+        dim: int, default=16
+            Dimension
+
+        num_axes: int, default=4
+            Number of axes
+
+        sigma_sq: dict, default=None
+            Sigma squared
+
+        sigma_sq_optimization: dict, default=None
+            Sigma squared optimization
+
+        seed: int, default=42
+            Seed
+
+        target_operator: list[str | int], default=None
+            Target operator
+
+        Returns
+        -------
+        AsyncQcogClient
+
+        """
+        super().ensemble(*args, **kwargs)
+        return self
+
+    # ###########################
+    # Public Utilities
+    # ###########################
+    def progress(self) -> dict:
+        """Return the current status of the training.
+
+        Returns
+        -------
+        dict
+            the current status of the training
+            `guid` : str
+            `training_completion` : int
+            `current_batch_completion` : int
+            `status` : TrainingStatus
+
+        """
+        return self.await_async(self._progress())
+
+    def status(self) -> TrainingStatus:
+        """Return the current status of the training.
+
+        Returns
+        -------
+        TrainingStatus
+            the current status of the training
+
+        """
+        return self.await_async(self._status())
+
+    def get_loss(self) -> Matrix | None:
+        """Return the loss matrix.
+
+        Returns
+        -------
+        Matrix
+            the loss matrix
+
+        """
+        return self.await_async(self._get_loss())
+
+    def wait_for_training(self, poll_time: int = 60) -> None:
+        """Wait for training to complete.
+
+        Note
+        ----
+        This function is blocking
+
+        Parameters
+        ----------
+        poll_time : int:
+            status checks intervals in seconds
+
+        Returns
+        -------
+        QcogClient
+            itself
+
+        """
+        self.await_async(self._wait_for_training(poll_time))
+
+    def await_async(
+        self, async_callable: Coroutine[Any, Any, CallableReturnType]
+    ) -> CallableReturnType:
+        """Await an async callable."""
+        # Make sure the function is not running in an event loop
+        try:
+            asyncio.get_running_loop()
+            # If a running event loop is present, we can't run async code
+            # in a synchronous function
+            raise RuntimeError("Cannot run async code in an event loop")
```

Comment:
with running loop it'll also end up in the `except` block below
```python
In [11]: asyncio.get_running_loop()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[11], line 1
----> 1 asyncio.get_running_loop()

RuntimeError: no running event loop
```

Reply:
oh fuck! ahahha yeah, I'm raising and excepting a RunTimeError!